Userland/Services/FileSystemAccessServer/ClientConnection.cpp

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2021, timmot <tiwwot@protonmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Debug.h>
+#include <FileSystemAccessServer/ClientConnection.h>
+#include <LibC/fcntl.h>
+#include <LibCore/File.h>
+#include <LibCore/StandardPaths.h>
+#include <LibGUI/Application.h>
+#include <LibGUI/FilePicker.h>
+
+namespace FileSystemAccessServer {
+
+static HashMap<int, RefPtr<ClientConnection>> s_connections;
+
+ClientConnection::ClientConnection(NonnullRefPtr<Core::LocalSocket> socket, int client_id)
+    : IPC::ClientConnection<FileSystemAccessClientEndpoint, FileSystemAccessServerEndpoint>(*this, move(socket), client_id)
+{
+    s_connections.set(client_id, *this);
+}
+
+ClientConnection::~ClientConnection()
+{
+}
+
+void ClientConnection::die()
+{
+    s_connections.remove(client_id());
+    exit(0);
+}
+
+Messages::FileSystemAccessServer::RequestFileReadResponse ClientConnection::request_file_read(String const& path)
+{
+    auto fd = open(path.characters(), O_RDONLY);
+
+    // TODO: Prompt user via new GUI control, the program is trying to access a file by itself.
+    // TODO: if root file...
+
+    return { -1, IPC::File(fd, IPC::File::CloseAfterSending) };
+}
+
+Messages::FileSystemAccessServer::PromptFileReadResponse ClientConnection::prompt_file_read(String const& path_to_view)
+{
+    auto app = GUI::Application::construct(0, nullptr);
+    auto user_picked_file = GUI::FilePicker::get_open_filepath(nullptr, "Select file", path_to_view);
+    app->exec();
+    app->quit();
+
+    if (user_picked_file.has_value()) {
+        auto fd = open(user_picked_file.value().characters(), O_RDONLY);
+
+        if (fd < 0) {
+            dbgln(OSError(errno).string());
+            return { -1, Optional<IPC::File> {} };
+        }
+
+        return { -1, IPC::File(fd, IPC::File::CloseAfterSending) };
+    }
+
+    return { -1, Optional<IPC::File> {} };
+}
+
+Messages::FileSystemAccessServer::RequestFileWriteResponse ClientConnection::request_file_write(String const& path)
+{
+    // TODO: Prompt user via new GUI control, the program is trying to access a file by itself.
+    // TODO: if root file...
+
+    (void)path;
+    return { -1, Optional<IPC::File> {} };
+}
+
+Messages::FileSystemAccessServer::PromptFileWriteResponse ClientConnection::prompt_file_write(String const& path_to_view)
+{
+    auto app = GUI::Application::construct(0, nullptr);
+    auto user_picked_file = GUI::FilePicker::get_save_filepath(nullptr, "save", "txt", path_to_view);
+    app->exec();
+    app->quit();
+
+    if (user_picked_file.has_value()) {
+        auto fd = open(user_picked_file.value().characters(), O_WRONLY | O_CREAT, 0644);
+
+        if (fd < 0) {
+            dbgln(OSError(errno).string());
+            return { -1, Optional<IPC::File> {} };
```

Comment:
Maybe you could change the IPC to return a `Result<IPC::File, int>`, which would allow you to e.g. use the positive numbers as the OS error codes, and the negative numbers for "own" errors, like:
* not an absolute path
* user terminated the dialog
* etc...

Reply:
Nice idea!

Reply:
Having more IPCCompiler issues when I try to use Result, I'm not sure I understand the error message entirely.


```cpp
// I've tried with and without the additional i32 error
request_file_read(String sender, String path) => (i32 error, Result<IPC::File,i32> fd_or_error)
```

```cpp
In file included from ../../Userland/Applications/Browser/Tab.cpp:41:
Userland/Services/FileSystemAccessServer/FileSystemAccessServerEndpoint.h: In static member function 'static AK::OwnPtr<Messages::FileSystemAccessServer::RequestFileReadResponse> Messages::FileSystemAccessServer::RequestFileReadResponse::decode(AK::InputMemoryStream&, int)':
Userland/Services/FileSystemAccessServer/FileSystemAccessServerEndpoint.h:69:46: error: could not convert '<brace-enclosed initializer list>()' from '<brace-enclosed initializer list>' to 'AK::Result<IPC::File, int>'
   69 |         Result<IPC::File,i32> fd_or_error = {};
      |                                              ^
      |                                              |
      |                                              <brace-enclosed initializer list>
In file included from Userland/Services/FileSystemAccessServer/FileSystemAccessClientEndpoint.h:13,
                 from ../../Userland/Applications/Browser/Tab.cpp:40:
../../Userland/Libraries/LibIPC/Decoder.h: In instantiation of 'bool IPC::decode(IPC::Decoder&, T&) [with T = AK::Result<IPC::File, int>]':
../../Userland/Libraries/LibIPC/Decoder.h:72:27:   required from 'bool IPC::Decoder::decode(T&) [with T = AK::Result<IPC::File, int>]'
Userland/Services/FileSystemAccessServer/FileSystemAccessServerEndpoint.h:70:28:   required from here
../../Userland/Libraries/LibIPC/Decoder.h:21:19: error: static assertion failed: Base IPC::decoder() instantiated
   21 |     static_assert(DependentFalse<T>, "Base IPC::decoder() instantiated");
      |                   ^~~~~~~~~~~~~~~~~
../../Userland/Libraries/LibIPC/Decoder.h:21:19: note: 'AK::Detail::DependentFalse<AK::Result<IPC::File, int> >' evaluates to false
In file included from Userland/Services/FileSystemAccessServer/FileSystemAccessClientEndpoint.h:15,
                 from ../../Userland/Applications/Browser/Tab.cpp:40:
../../Userland/Libraries/LibIPC/Encoder.h: In instantiation of 'bool IPC::encode(IPC::Encoder&, T&) [with T = const AK::Result<IPC::File, int>]':
../../Userland/Libraries/LibIPC/Encoder.h:84:20:   required from 'void IPC::Encoder::encode(const T&) [with T = AK::Result<IPC::File, int>]'
../../Userland/Libraries/LibIPC/Encoder.h:68:15:   required from 'IPC::Encoder& IPC::Encoder::operator<<(const T&) [with T = AK::Result<IPC::File, int>]'
Userland/Services/FileSystemAccessServer/FileSystemAccessServerEndpoint.h:89:19:   required from here
../../Userland/Libraries/LibIPC/Encoder.h:17:19: error: static assertion failed: Base IPC::encode() was instantiated
   17 |     static_assert(DependentFalse<T>, "Base IPC::encode() was instantiated");
      |                   ^~~~~~~~~~~~~~~~~
../../Userland/Libraries/LibIPC/Encoder.h:17:19: note: 'AK::Detail::DependentFalse<const AK::Result<IPC::File, int> >' evaluates to false
```

Reply:
I'm going to resolve this, but I'll keep your ideas about error numbers in mind.